File: sdks/dotnet/Nivrit/Session.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Nivrit;

public class NivritSession
{
    public NivritClient Client { get; }
    public HelperCrypto Crypto { get; }
    public JsonNode User { get; private set; } = new JsonObject();
    public string PrivateKey { get; private set; } = "";
    private readonly Dictionary<string, string> _projectKeys = new();

    public NivritSession(string baseUrl, string token, HelperCrypto crypto)
    {
        Client = new NivritClient(baseUrl, token);
        Crypto = crypto;
    }

    public async Task AuthenticateAsync(string password)
    {
        User = await Client.GetMeAsync();
        PrivateKey = Crypto.DecryptPrivateKey(
            (string?)User["encrypted_private_key"] ?? throw new InvalidOperationException("missing encrypted_private_key"),
            (string?)User["private_key_nonce"] ?? throw new InvalidOperationException("missing private_key_nonce"),
            password);
    }

    public async Task<IReadOnlyList<JsonObject>> ListProjectsAsync(string orgId)
    {
        var projects = await Client.ListOrgProjectsAsync(orgId);
        var memberships = await Client.ListMyProjectsAsync();
        var membershipMap = new Dictionary<string, JsonNode>();
        foreach (var m in memberships!.AsArray())
        {
            var projectId = (string?)m?["project_id"]
                ?? throw new InvalidOperationException("membership missing project_id");
            membershipMap[projectId] = m!;
        }

        var result = new List<JsonObject>();
        foreach (var p in projects!.AsArray())
        {
            var project = p?.AsObject()
                ?? throw new InvalidOperationException("invalid project");
            var projectId = (string?)project["id"]
                ?? throw new InvalidOperationException("project missing id");
            var obj = (JsonObject)project.DeepClone();
            obj["membership"] = membershipMap.TryGetValue(projectId, out var mem) ? mem : null;
            result.Add(obj);
        }
        return result;
    }

    public string GetProjectKey(JsonNode membership)
    {
        var pid = (string?)membership["project_id"] ?? throw new InvalidOperationException("missing project_id");
        if (_projectKeys.TryGetValue(pid, out var cached)) return cached;
        var key = Crypto.DecapsulateProjectKey(
            (string?)membership["encrypted_project_key"] ?? throw new InvalidOperationException("missing encrypted_project_key"),
            PrivateKey);
        _projectKeys[pid] = key;
        return key;
    }

    public async Task<IReadOnlyList<JsonObject>> ListSecretsAsync(string projectId, string environmentId)
    {
        var secrets = await Client.ListSecretsAsync(projectId, environmentId);
        var memberships = await Client.ListMyProjectsAsync();
        JsonNode? membership = null;
        foreach (var m in memberships!.AsArray())
            if ((string?)m!["project_id"] == projectId) { membership = m; break; }
        if (membership == null) throw new InvalidOperationException($"No membership for project {projectId}");
        var projectKey = GetProjectKey(membership);

        var result = new List<JsonObject>();
        foreach (var s in secrets!.AsArray())
        {
            var secret = s?.AsObject()
                ?? throw new InvalidOperationException("invalid secret");
            var obj = (JsonObject)secret.DeepClone();
            obj["value"] = Crypto.DecryptValue(
                (string?)secret["encrypted_value"]
                    ?? throw new InvalidOperationException("secret missing encrypted_value"),
                (string?)secret["nonce"]
                    ?? throw new InvalidOperationException("secret missing nonce"),
                projectKey);
            result.Add(obj);
        }
        return result;
    }

    public static string Base64Encode(byte[] data) => Convert.ToBase64String(data);
    public static byte[] Base64Decode(string s) => Convert.FromBase64String(s);
}

File: sdks/elixir/lib/nivrit/session.ex
defmodule Nivrit.Session do
  @moduledoc "Authenticated session with cached decrypted project keys."

  defstruct [:client, :crypto, :user, :private_key, project_keys: %{}]

  def new(base_url, token, crypto) do
    %__MODULE__{client: Nivrit.Client.new(base_url, token), crypto: crypto}
  end

  def authenticate(%__MODULE__{} = session, password) do
    user = Nivrit.Client.get_me(session.client)

    private_key =
      Nivrit.Crypto.decrypt_private_key(
        session.crypto,
        user["encrypted_private_key"],
        user["private_key_nonce"],
        password
      )

    %{session | user: user, private_key: private_key}
  end

  def list_projects(%__MODULE__{} = session, org_id) do
    projects = Nivrit.Client.list_org_projects(session.client, org_id)
    memberships = Nivrit.Client.list_my_projects(session.client) |> Map.new(&{&1["project_id"], &1})
    Enum.map(projects, &Map.put(&1, "membership", memberships[&1["id"]]))
  end

  def get_project_key(%__MODULE__{} = session, membership) do
    pid = membership["project_id"]

    case session.project_keys[pid] do
      nil ->
        key =
          Nivrit.Crypto.decapsulate_project_key(
            session.crypto,
            membership["encrypted_project_key"],
            session.private_key
          )

        %{session | project_keys: Map.put(session.project_keys, pid, key)}
        key

      key ->
        key
    end
  end

  def list_secrets(%__MODULE__{} = session, project_id, environment_id) do
    secrets = Nivrit.Client.list_secrets(session.client, project_id, environment_id)
    membership = Enum.find(Nivrit.Client.list_my_projects(session.client), &(&1["project_id"] == project_id))
    project_key = get_project_key(session, membership)

    Enum.map(secrets, fn s ->
      value = Nivrit.Crypto.decrypt_value(session.crypto, s["encrypted_value"], s["nonce"], project_key)
      Map.put(s, "value", value)
    end)
  end
end
